### ISSUE TITLE
[FEATURE] - Create a Command to Index Existing Tests for Semantic Search

### ISSUE THREAD

**Issue by Ofido** [Ofido/AIUnitTest#8]
### Is your feature request related to a problem?

To enable semantic search, the existing test suite needs to be processed and indexed into a searchable format (e.g., a vector database or a local file). This process should be user-initiated and managed through a dedicated command.

### Describe the solution you'd like

Create a new CLI command, for example `ai-unit-test index`, that performs the following actions:
1.  Finds all existing test files in the project.
2.  Reads the content of each test file.
3.  Chunks the tests into logical blocks (e.g., per function or class).
4.  Generates vector embeddings for each chunk.
5.  Saves these embeddings to a local file (e.g., `.ai-unit-test-cache/index.faiss`) that can be used by the semantic search feature.

This command allows the user to control when the indexing happens and provides a clear separation of concerns.

### Describe alternatives you've considered

- Indexing automatically in the background (less transparent for the user).
- Indexing on-the-fly with every run (inefficient and slow).

### Additional context

This is a prerequisite for the **Semantic Search (Embeddings)** feature and is part of **Phase 2: Advanced Test Generation (RAG)**.

**Comment by Ofido** [comment]
Hi @onestardao 

I’m thinking about different integration paths for WFGY in #26 , and I’d really value your perspective. Here are the options I see right now:

- Proxy Architecture – Intercept LLM calls with a wrapper that simulates logits.
- Embedding-Space WFGY – Adapt WFGY to operate on embeddings or semantic outputs instead of raw logits.
- Multi-Model Ensemble – Use WFGY with a local model as a validator for API outputs.
- Semantic Consistency Layer – Apply WFGY-like principles on the semantic space of generated tests, ensuring consistency across runs.
- Hybrid Pipeline – Combine APIs and local models strategically (e.g., APIs for speed, local models + WFGY for validation).

but in mind my main goal with AIUnitTest is to make it useful for as many people as possible. I don’t want the project to be limited only to developers who can run local models with logits access — I’d like to support both commercial APIs (e.g., OpenAI, Anthropic) and local/self-hosted models where logits are available.

if u can help me, from your perspective, would simply adjusting the prompt be enough to apply WFGY’s principles in the API-based workflow? Or do you think a deeper integration is always required?

Thanks a lot for your time and insights 🙏

(PS: I haven't started integrating anything yet and I think the code I made is messy so it might still need a little reorganization before I start with that)

**Comment by onestardao** [comment]
thanks for laying out these paths so clearly  this is actually the exact tradeoff spectrum we’ve seen with other teams too.
to answer your question:

prompt-level WFGY works as a lightweight semantic firewall, and it already fixes ~60–70% of the recurring drift modes without touching infra.

deeper integration (logit or embedding layer) is where you unlock the last 20–30% 

consistency across runs, reproducibility, and coverage guarantees.

so the pattern we usually recommend:

start with prompt-level, validate that ΔS ≤ 0.45 and coverage ≥ 0.70.

once stable, decide whether to invest in logit/embedding hooks for long-term reproducibility.

happy to help you sketch the minimal patch first, and we can evolve toward the deeper path once it proves useful in your pipeline.

**Comment by Ofido** [comment]
Thanks a lot for breaking this down so clearly — the ΔS ≤ 0.45 / coverage ≥ 0.70 targets are super helpful as a first checkpoint.

I think starting with the prompt-level WFGY as a lightweight semantic firewall makes the most sense for AIUnitTest right now, since it keeps the project broadly usable (even for folks who rely only on API-based models). Once we have stability and a clear view of the improvement, we can explore the embedding/logit-level path for reproducibility and long-term robustness.

If you’re open, I’d love to collaborate on sketching that minimal patch you mentioned. It would be great to align on what the smallest viable integration looks like, so that we can merge quickly and then iterate.

Thanks again for sharing your experience — this is exactly the kind of guidance that helps keep the project practical while still moving toward state-of-the-art methods. 🙏

**Comment by onestardao** [comment]
**Plan: Minimal Integration MVP for WFGY semantic firewall**

thanks for the green light. i’ll push a tiny MVP so we can validate the idea in your stack first, then decide how deep we go.

### 0) goal in one line

add a lightweight **pre-gen semantic firewall** that checks role drift and context fit before the agent/tool runs. pass if ΔS ≤ 0.45 and coverage ≥ 0.70. if not, auto-reset or compress.

### 1) insertion point for MVP

**Path A: prompt-wrapper proxy** (no infra change)

* intercept request right before model call
* compute ΔS with anchors
* if drift: rewrite or re-init, else pass through

we can add Path B (embedding gate) later if you want, but path A gets us a fast win.

### 2) deliverables

* `wfgy_guard.py`  a single file CLI
* `spec.yaml`      anchors + weights + thresholds
* `examples/`      3 tiny tasks for repro
* `report.json`    per call telemetry

### 3) CLI

```bash
# init a sample spec
python wfgy_guard.py init --out spec.yaml

# run guard once
python wfgy_guard.py guard \
  --input input.txt \
  --goal goal.txt \
  --spec spec.yaml \
  --out report.json

# pipe mode for easy proxy
cat prompt.txt | python wfgy_guard.py guard --goal goal.txt --spec spec.yaml
```

### 4) spec.yaml (minimal)

```yaml
anchors:
  entities: ["invoice", "due date", "customer id"]
  relations: ["invoice->due date", "customer id->lookup"]
  constraints: ["net-30", "usd only"]
weights:
  w_e: 0.5
  w_r: 0.3
  w_c: 0.2
thresholds:
  delta_s_max: 0.45
  coverage_min: 0.70
actions:
  on_drift: ["reset_role", "compress_context"]
  on_low_cov: ["add_missing_anchors"]
```

### 5) telemetry (always printed)

```
triage=No.1|No.5  delta_s=0.38  coverage=0.74  lambda=convergent  decision=pass
```

* `triage` maps to failure patterns we observe during guard
* `decision` is `pass|rewrite|reset`

### 6) acceptance targets

* ΔS ≤ 0.45 on 3 consecutive runs for the same task
* coverage ≥ 0.70 with anchors present
* no role drift across retry 1→3

### 7) quick dataset to reproduce

* `ex1_tool_pick.txt`  tests wrong tool loop
* `ex2_role_swap.txt`  tests role drift
* `ex3_cold_boot.txt`  tests first call garbage

### 8) scope and time

* day 1  ship CLI + spec + examples
* day 2  wire into your pipeline as a thin proxy function
* day 3  tune anchors with your real tasks, confirm targets
* day 4  optional: add `--ci` mode to fail builds on drift

### 9) extension options after MVP

* Path B: **embedding gate** before retrieval, normalize vectors, add trace IDs
* Path C: **multi-agent lock ordering** and token passing to prevent deadlock
* Path D: **recovery bridges** for stalled steps

if this looks good, i’ll open a PR with the single-file CLI and the sample spec, then we tune it together on your tasks. happy to adjust the insertion point if you prefer embedding gate first. waiting on your preference and any constraints i should match.
